dll/win32/shdocvw/CNSCBand.cpp

```diff
@@ -0,0 +1,1682 @@
+        if (FAILED_UNEXPECTEDLY(hr))
+            return hr;
+    }
+
+    return (*ppFolder)->EnumObjects(NULL, _GetEnumFlags(), &pEnum);
+}
+
+BOOL CNSCBand::_ItemHasAnyChild(_In_ HTREEITEM hItem)
+{
+    CComPtr<IEnumIDList> pEnum;
+    HRESULT hr = _GetItemEnum(pEnum, hItem);
+    if (FAILED(hr))
+        return FALSE;
+
+    CComHeapPtr<ITEMIDLIST_RELATIVE> pidlTemp;
+    hr = pEnum->Next(1, &pidlTemp, NULL);
+    return SUCCEEDED(hr);
+}
+
+void CNSCBand::_RefreshRecurse(_In_ HTREEITEM hTarget)
+{
+    CComPtr<IEnumIDList> pEnum;
+    HRESULT hrEnum = _GetItemEnum(pEnum, hTarget);
+
+    // Delete zombie items
+    HTREEITEM hItem, hNextItem;
+    for (hItem = TreeView_GetChild(m_hwndTreeView, hTarget); hItem; hItem = hNextItem)
+    {
+        hNextItem = TreeView_GetNextSibling(m_hwndTreeView, hItem);
+
+        if (SUCCEEDED(hrEnum) && !_IsTreeItemInEnum(hItem, pEnum))
+            TreeView_DeleteItem(m_hwndTreeView, hItem);
+    }
+
+    pEnum = NULL;
+    hrEnum = _GetItemEnum(pEnum, hTarget);
+
+    CItemData* pItemData = ((hTarget == TVI_ROOT) ? NULL : GetItemData(hTarget));
+
+    // Insert new items and update items
+    if (SUCCEEDED(hrEnum))
+    {
+        CComHeapPtr<ITEMIDLIST_RELATIVE> pidlTemp;
+        while (pEnum->Next(1, &pidlTemp, NULL) == S_OK)
+        {
+            if (!_TreeItemHasThisChild(hTarget, pidlTemp))
+            {
+                if (pItemData)
+                {
+                    CComHeapPtr<ITEMIDLIST> pidlAbsolute(ILCombine(pItemData->absolutePidl, pidlTemp));
+                    _InsertItem(hTarget, pidlAbsolute, pidlTemp, TRUE);
+                }
+                else
+                {
+                    CComHeapPtr<ITEMIDLIST> pidlAbsolute(ILCombine(m_pidlRoot, pidlTemp));
+                    _InsertItem(hTarget, pidlAbsolute, pidlTemp, TRUE);
+                }
+            }
+            pidlTemp.Free();
+        }
+    }
+
+    // Update children and recurse
+    for (hItem = TreeView_GetChild(m_hwndTreeView, hTarget); hItem; hItem = hNextItem)
+    {
+        hNextItem = TreeView_GetNextSibling(m_hwndTreeView, hItem);
+
+        TV_ITEMW item = { TVIF_HANDLE | TVIF_CHILDREN };
+        item.hItem = hItem;
+        item.cChildren = _ItemHasAnyChild(hItem);
+        TreeView_SetItem(m_hwndTreeView, &item);
+
+        if (TreeView_GetItemState(m_hwndTreeView, hItem, TVIS_EXPANDEDONCE) & TVIS_EXPANDEDONCE)
+            _RefreshRecurse(hItem);
+    }
+}
+
+void CNSCBand::_Refresh()
+{
+    m_hwndTreeView.SendMessage(WM_SETREDRAW, FALSE, 0);
+    _RefreshRecurse(_WantsRootItem() ? m_hRoot : TVI_ROOT);
+    m_hwndTreeView.SendMessage(WM_SETREDRAW, TRUE, 0);
+}
+
+LRESULT CNSCBand::OnTimer(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL &bHandled)
+{
+    KillTimer(wParam);
+
+    switch (wParam)
+    {
+        case TIMER_ID_REFRESH:
+            _Refresh();
+            break;
+    }
```

Comment:
Currently we could use a simple if here instead of the switch. But whether that would be a good idea ofc depends on whether the switch would be extended later. For now that might mute some static-code-analysis-tools though, which are sometimes greedy to always have a default-case.

Reply:
Fixed.